File: src/day12.ts
export function solve_day_12(input: string): number {
    // Parse the input to create the grid
    const grid: string[][] = input.trim().split("\n").map(line => line.trim().split(""));

    const price = calculateTotalPrice(grid);

    console.log(`Prix : ${price}`);
    return price;
}


type Position = { x: number; y: number };

function calculateTotalPrice(grid: string[][]): number {
    const rows: number = grid.length;
    const cols: number = grid[0].length;
    const visited: boolean[][] = Array(rows).fill(0).map(() => Array(cols).fill(false));

    const directions: Position[] = [
        { x: -1, y: 0 },
        { x: 1, y: 0 },
        { x: 0, y: -1 },
        { x: 0, y: 1 },
    ];

    const isInsideGrid = (x: number, y: number): boolean => {
        return x >= 0 && x < rows && y >= 0 && y < cols;
    }


    function bfs(start: Position, plantType: string): { area: number; sides: number } {
        let queue: Position[] = [start];
        visited[start.x][start.y] = true;

        let area = 0;
        let sides = 0;

        while (queue.length > 0) {
            const { x, y } = queue.shift()!;
            area++;

            for (const dir of directions) {
                const nx = x + dir.x;
                const ny = y + dir.y;

                // Vérifie si le côté est un bord ou une séparation
                if (!isInsideGrid(nx, ny) || grid[nx][ny] !== plantType) {
                    sides++;
                } else if (!visited[nx][ny]) {
                    // Ajoute la cellule voisine à la file pour exploration
                    visited[nx][ny] = true;
                    queue.push({ x: nx, y: ny });
                }
            }
        }

        return { area, sides };
    }

    let totalPrice = 0;

    for (let r = 0; r < rows; r++) {
        for (let c = 0; c < cols; c++) {
            if (!visited[r][c]) {
                const plantType = grid[r][c];
                const { area, sides } = bfs({ x: r, y: c }, plantType);
                totalPrice += area * sides;
            }
        }
    }

    return totalPrice;
}

File: srcr/src/days/day12.rs
use std::collections::VecDeque;

pub fn solve_day12(input: &str) -> u32 {

    let mut grid: Vec<Vec<char>> = input.trim().split("\n")
        .map(|line| line.trim().chars().collect()).collect();

    let price = calculate_total_price(&mut grid);

    println!("Prix : {}", price);
    return price;
}


fn calculate_total_price(grid: &mut Vec<Vec<char>>) -> u32 {
    let rows: usize = grid.len();
    let cols: usize = grid[0].len();
    let mut visited: Vec<Vec<bool>> = vec![vec![false; cols]; rows];

    let mut total_price = 0;

    for r in 0..rows {
        for c in 0..cols {
            if !visited[r][c] {
                let plant_type = grid[r][c];
                let (area, sides) = bfs(Position { x: r, y: c }, plant_type, &mut visited, &grid);
                total_price += area * sides;
            }
        }
    }

    total_price
}


struct Position {
    x: usize,
    y: usize,
}

fn bfs(
    start: Position,
    plant_type: char,
    visited: &mut Vec<Vec<bool>>,
    grid: &Vec<Vec<char>>
) -> (u32, u32)
{
    let mut queue = VecDeque::new();
    visited[*&start.x][*&start.y] = true;
    queue.push_back(start);

    let mut area = 0;
    let mut sides = 0;

    let directions: Vec<Position> = vec![
        Position { x: 0, y: 1 },
        Position { x: 1, y: 0 },
        Position { x: 0, y: usize::MAX },
        Position { x: usize::MAX, y: 0 },
    ];

    while let Some(Position { x, y }) = queue.pop_front() {
        area += 1;

        for dir in &directions {
            let nx = if dir.x == usize::MAX { x.wrapping_sub(1) } else { x + dir.x };
            let ny = if dir.y == usize::MAX { y.wrapping_sub(1) } else { y + dir.y };

            if nx >= grid.len() || ny >= grid[0].len() || grid[nx][ny] != plant_type {
                sides += 1;
            } else if !visited[nx][ny] {
                visited[nx][ny] = true;
                queue.push_back(Position { x: nx, y: ny });
            }
        }
    }
    (area, sides)
}